### PR TITLE
[ADD] hr_expense_cancel

### PR DESCRIPTION
**Pull request by JordiBForgeFlow**
This module allows you to:
- Cancel an Expense that is in status 'Draft'
- Cancel and revert back to 'Draft' an Expense, in state 'Waiting
  Payment'.
- Revert back to 'Waiting Payment' an Expense that has been paid, if the
  associated payment is unreconciled.

### CODE REVIEW

File: hr_expense_cancel/models/hr_expense.py [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+            expense.payment_ids = lines
+
+    payment_ids = fields.Many2many(compute=_compute_payment_ids,
+                                   comodel_name='account.move.line',
+                                   string='Payments')
+
+    @api.multi
+    def expense_canceled(self):
+        account_move_obj = self.env['account.move']
+        moves = account_move_obj.browse([])
```
Comment (ghost) [inline hr_expense_cancel/models/hr_expense.py:44]: if you browse an empty record set, 

moves = self.env['account.move']

is enough

File: hr_expense_cancel/models/hr_expense.py [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+        for expense in self:
+            if expense.account_move_id:
+                moves += expense.account_move_id
```
Comment (ghost) [inline hr_expense_cancel/models/hr_expense.py:47]: wrong operator, to join recordsets, use

moves |= expense.account_move_id

File: hr_expense_cancel/models/hr_expense.py [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+            if expense.payment_ids:
```
Comment (ghost) [inline hr_expense_cancel/models/hr_expense.py:48]: your if statement is redondant

File: hr_expense_cancel/models/hr_expense.py [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+                        temp_lines = map(
+                            lambda x1: x, m.reconcile_id.line_id)
+                    elif m.reconcile_partial_id:
+                        temp_lines = map(
+                            lambda x2: x,
+                            m.reconcile_partial_id.line_partial_ids)
+                    lines += [x for x in temp_lines if x not in lines]
+                    src.append(m)
+
+            lines = filter(lambda x3: x not in src, lines)
```
Comment (ghost) [inline hr_expense_cancel/models/hr_expense.py:34]: filter and map instructions return lists. better work with recordsets all the way

File: hr_expense_cancel/views/hr_expense_view.xml [hunk 5]
```diff
@@ -0,0 +1,17 @@
...
+        <record id="view_expenses_form_ext" model="ir.ui.view">
+            <field name="name">view.expenses.form.ext</field>
+            <field name="inherit_id" ref="hr_expense.view_expenses_form" />
+            <field name="model">hr.expense.expense</field>
+            <field name="arch" type="xml">
+                <data>                    
+                    <xpath expr="/form/header/button[@name='refuse']" position="after">
+                        <button name="cancel" states="draft,done" string="Cancel" type="workflow" groups="base.group_hr_user" />
+                    </xpath>	
+				</data>
```
Comment (ghost) [inline hr_expense_cancel/views/hr_expense_view.xml:13]: extra indent

File: hr_expense_cancel/views/hr_expense_view.xml [hunk 6]
```diff
@@ -0,0 +1,17 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<openerp>
+    <data>
+        <record id="view_expenses_form_ext" model="ir.ui.view">
+            <field name="name">view.expenses.form.ext</field>
+            <field name="inherit_id" ref="hr_expense.view_expenses_form" />
+            <field name="model">hr.expense.expense</field>
+            <field name="arch" type="xml">
+                <data>                    
+                    <xpath expr="/form/header/button[@name='refuse']" position="after">
```
Comment (ghost) [inline hr_expense_cancel/views/hr_expense_view.xml:10]: can be simplified to

//button[@name='refuse']

File: hr_expense_cancel/views/hr_expense_view.xml [hunk 7]
```diff
@@ -0,0 +1,17 @@
...
+                        <button name="cancel" states="draft,done" string="Cancel" type="workflow" groups="base.group_hr_user" />
```
Comment (ghost) [inline hr_expense_cancel/views/hr_expense_view.xml:11]: you might have an issue with security. I don't think group_hr_user have the sufficient access on account.move and account.move.line

**Comment by pedrobaeza** [comment]
There's already a PR for a module like this: https://github.com/OCA/hr/pull/202

**Comment by feketemihai** [comment]
Please reopen the PR if it's important to you.

**Pull request closed without merging**